index.js

```diff
@@ -220,18 +234,28 @@ class TuyaDevice extends EventEmitter {
       sequenceN: ++this._currentSequenceN
     });
 
-    // Send request and wait for response
-    this._waitingForSetToResolve = true;
-    return new Promise((resolve, reject) => {
+    // Queue this request and limit concurrent set requests to one
+    return this._setQueue.add(() => new Promise((resolve, reject) => {
       try {
-        // Send request
-        this._send(buffer);
-
-        this._setResolver = resolve;
+        const data = pTimeout(new Promise((resolve, reject) => {
+          // Send request and wait for response
+          try {
+            // Send request
+            this._send(buffer);
+            this._setResolver = resolve;
+          } catch (error) {
+            reject(error);
+          }
+        }), this._responseTimeout * 1000, () => {
+          // Only gets here on timeout so clear resolver function and emit error
+          this._setResolver = undefined;
+          this.emit('error', 'Timeout waiting for status response from device id: ' + this.device.id);
+        });
+        resolve(data);
```

Comment:
I'm a little confused on how this works. Isn't `data` when `resolve()` called a Promise, not the result?

I think the top-level Promise (`this._setQueue.add(() => new Promise((resolve, reject) =>`) can be changed to an async arrow function to simplify some of this.

Reply:
This is most likely due to my limited Javascript abilities.  I feel like I've got a good handle on what the code is doing, but my skillset with Javascript is still quite basic as I'm just learning as I hack on these projects.

I originally did have an async arrow function there, but 'npm test' said this wasn't allowed (I don't remember the exact message but something about promise should not use async function?)

I originally tried to do this without the assignment to data at all, just return pTimeout (which is what happens in find) but somehow I could never get the syntax and behavior to work correctly.  I'll take another stab at it as I agree it's hard to follow.

Reply:
Updated with a new version that removes the extra, unnecessary promise.  Not sure why that was so hard for me the first time, seemed super clear today.  Let me know if you think this is better or could still use improvement.

Reply:
Looks great!